packages/ra-core/src/core/CoreAdminContext.tsx

```diff
@@ -65,14 +65,33 @@ React-admin requires a valid dataProvider function to work.`);
             ? convertLegacyDataProvider(dataProvider)
             : dataProvider;
 
-    const finalHistory = history || createHashHistory();
+    let historyRef = React.useRef<History>();
+    if (historyRef.current == null) {
+        historyRef.current = createHashHistory({ window });
+    }
+
+    let finalHistory = historyRef.current;
+    let [historyState, setHistoryState] = React.useState({
+        action: finalHistory.action,
+        location: finalHistory.location,
+    });
+
+    React.useLayoutEffect(() => finalHistory.listen(setHistoryState), [
```

Comment:
I don't understand why you need that state. Can't you pass only the history?

Reply:
I just copied the code from react-router. See https://github.com/remix-run/react-router/blob/main/packages/react-router-dom/index.tsx#L133 & https://github.com/remix-run/react-router/blob/main/packages/react-router-dom/index.tsx#L172